### PR TITLE
Embedding list enum

### PR DESCRIPTION
**Pull request by JWittmeyer**
- https://github.com/code-kern-ai/refinery-gateway/pull/148
- https://github.com/code-kern-ai/refinery-submodule-model/pull/51
- https://github.com/code-kern-ai/refinery-ac-exec-env/pull/27
- https://github.com/code-kern-ai/refinery-embedder/pull/58
- https://github.com/code-kern-ai/refinery-ui/pull/153
- https://github.com/code-kern-ai/refinery-neural-search/pull/36
- https://github.com/code-kern-ai/gates-runtime/pull/31

Note:
Gates was pretty much left untouched for the moment since other prs that change the structure are still in review queue & I'm unsure how we want to approach gates in the future (specify what is wanted, e.g. the new embedding_list attribute is atm fully returned etc,)

Best option to test:
`sudo bash start -g -b embedding-list` (newest dev setup version is needed -> this will start refinery & gates on the specified branch)

**otherwise:** remember to build the ac exec env & gates runtime to test :)

### CODE REVIEW

File: business_objects/embedding.py [hunk 1]
```diff
@@ -200,12 +201,20 @@ def get_tensors_and_attributes_for_qdrant(
             if payload_selector != "":
                 payload_selector += ","
             if data_type != enums.DataTypes.TEXT.value:
-                payload_selector += f"'{attr}', (r.\"data\"->>'{attr}')::{data_type}"
+                payload_selector += f"'{attr}', (r.\"data\"->'{attr}')::{data_type}"
```
Comment (jhoetter) [inline business_objects/embedding.py:204]: minor hint: the difference between -> and ->> are not clear to me. Again, just for the dumber programmers (me), the comment why the change is relevant (or even _that_ there is a change) would have been cool for me :-)

File: business_objects/embedding.py [hunk 2]
```diff
@@ -215,6 +224,39 @@ def get_tensors_and_attributes_for_qdrant(
...
+) -> int:
+    query = f"""
+    SELECT CEIL(AVG(max_key))::INTEGER
+    FROM (
+        SELECT record_id, MAX(sub_key + 1) max_key
+        FROM embedding_tensor et
+        WHERE project_id = '{project_id}' AND embedding_id = '{embedding_id}'
+        GROUP BY record_id )x """
+    value = general.execute_first(query)
+    if value is None or value[0] is None:
```
Comment (jhoetter) [inline business_objects/embedding.py:255]: not sure: can it be that value is `[]`? or is that never the case? if it can be, a check for list length could be reasonable. If it can't be, just resolve :-)

File: business_objects/embedding.py [hunk 3]
```diff
@@ -355,16 +419,30 @@ def create_tensors(
...
-            project_id=project_id,
-            record_id=record_id,
-            embedding_id=embedding_id,
-            data=tensor,
-        )
-        for record_id, tensor in zip(record_ids, tensors)
-    ]
-    general.add_all(tensors, with_commit)
+    to_add = None
+    if len(record_ids) > 0 and "@" in record_ids[0]:
```
Comment (jhoetter) [inline business_objects/embedding.py:423]: what is the meaning of the "@" :D (sorry if I'm adding all the comments regarding readability, but to me it isn't clear what the context is directly)

Comment (jhoetter) [reply business_objects/embedding.py:423]: is it meant as something to split the text?

**Review by jhoetter — APPROVED** [review]
only few comments regarding documentation, looks good to me and works

File: business_objects/embedding.py (hunk 1 above)
Comment (JWittmeyer) [reply business_objects/embedding.py:204]: in short previous version was faulty. -> returns a json object ->> directly casts to string. Since we cast after that anyway the second > was unnecessary from the initial introduction

Comment (JWittmeyer) [reply business_objects/embedding.py:204]: also i'm not sure if we should start explaining operators in comments because there will be no end to it

File: business_objects/embedding.py (hunk 2 above)
Comment (JWittmeyer) [reply business_objects/embedding.py:255]: value is a scalar through execute_first. the [0] ist access of the first entry first columvalue so -> can't be an empty list

File: business_objects/embedding.py (hunk 3 above)
Comment (JWittmeyer) [reply business_objects/embedding.py:423]: added comment & yes
